cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +562,514 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
+        return true;
+    }
+    if (diff > 0) { // r1 start is strictly after r2 start.
+        return false;
+    }
+    const auto len1 = r1.start()->value().representation().size();
+    const auto len2 = r2.start()->value().representation().size();
+    if (len1 == len2) { // The values truly are equal.
+        return r1.start()->is_inclusive() && !r2.start()->is_inclusive();
+    } else if (len1 < len2) { // r1 start is a prefix of r2 start.
+        // (a)>=(1) starts before (a,b)>=(1,1), but (a)>(1) doesn't.
+        return r1.start()->is_inclusive();
+    } else { // r2 start is a prefix of r1 start.
+        // (a,b)>=(1,1) starts before (a)>(1) but after (a)>=(1).
+        return r2.start()->is_inclusive();
+    }
+}
+
+/// True iff r1 start is before r2 end.
+bool starts_before_end(
+        const query::clustering_range& r1,
+        const query::clustering_range& r2,
+        const clustering_key_prefix::prefix_equal_tri_compare& cmp) {
+    if (!r1.start()) {
+        return true; // r1 start is -inf, must be before r2 end.
+    }
+    if (!r2.end()) {
+        return true; // r2 end is +inf, everything is before it.
+    }
+    const auto diff = cmp(r1.start()->value(), r2.end()->value());
+    if (diff < 0) { // r1 start is strictly before r2 end.
+        return true;
+    }
+    if (diff > 0) { // r1 start is strictly after r2 end.
+        return false;
+    }
+    const auto len1 = r1.start()->value().representation().size();
+    const auto len2 = r2.end()->value().representation().size();
+    if (len1 == len2) { // The values truly are equal.
+        return r1.start()->is_inclusive() && r2.end()->is_inclusive();
+    } else if (len1 < len2) { // r1 start is a prefix of r2 end.
+        // a>=(1) starts before (a,b)<=(1,1) ends, but (a)>(1) doesn't.
+        return r1.start()->is_inclusive();
+    } else { // r2 end is a prefix of r1 start.
+        // (a,b)>=(1,1) starts before (a)<=(1) ends but after (a)<(1) ends.
+        return r2.end()->is_inclusive();
+    }
+}
+
+/// True if r1 end is before r2 end.
+bool ends_before_end(
+        const query::clustering_range& r1,
+        const query::clustering_range& r2,
+        const clustering_key_prefix::prefix_equal_tri_compare& cmp) {
+    if (!r1.end()) {
+        return false; // r1 end is +inf, which is after everything.
+    }
+    if (!r2.end()) {
+        return true; // r2 end is +inf, while r1 end is finite.
+    }
+    const auto diff = cmp(r1.end()->value(), r2.end()->value());
+    if (diff < 0) { // r1 end is strictly before r2 end.
+        return true;
+    }
+    if (diff > 0) { // r1 end is strictly after r2 end.
+        return false;
+    }
+    const auto len1 = r1.end()->value().representation().size();
+    const auto len2 = r2.end()->value().representation().size();
+    if (len1 == len2) { // The values truly are equal.
+        return !r1.end()->is_inclusive() || r2.end()->is_inclusive();
```

Comment:
What if both are inclusive? Shouldn't we require that r1's end is exclusive and r2's is inclusive?

Reply:
You're right, fixed.  I also specified explicitly when "before" means strictly before and when it means "before or at".